### DISCUSSION TITLE
Periodic Orchard Core workshops

### DISCUSSION THREAD

**Discussion by Piedone** [OrchardCMS/OrchardCore#14449] [Ideas]
In 2020, as a kind of substitute to Harvest (that was meant to be organized that year in person but was called off due to COVID), we had Orchard Core workshops, see https://orchardcore.net/workshops.

I think we should have such workshops (beginner user/admin area, beginner developer/theming, intermediate developer/module development, Orchard Core Commerce, deploying to Azure) periodically, like every quarter. This would maintain an accessible version of Orchard Core training, making Orchard Core itself more accessible. These would again be paid, like $200 a piece, or $150 if you've ever contributed to OC, and provided either by willing community members. We at Lombiq can provide several too.

What do you think? Especially @agriffard @bleroy @deanmarcussen @domonkosgabor since you've given workshops. The Orchard Core Commerce one this time would perhaps be by @DAud-IcI.

**Comment by hishamco** [comment]
I totally agree to have such workshops annually if it's possible. Part of that we can have some FREE as a contribution to the community and we could look for different ways to sponsor them

**Reply by Piedone** [reply]
If the people giving the workshops can offer that, we could have 1-2 free spots, for example. But such a workshop is quite a lot of work to prepare and hold, and it's only for the people attending (as opposed to the wider community) so it's adequate that it's paid.

**Comment by duxonline** [comment]
This was exactly what I was looking for when starting with OrchardCore. I'm still interested in intermediate module/theme development training. Time zone can be an issue for us in Australia. Hopefully this will be taken into consideration.

**Reply by Piedone** [reply]
Great then! I'm sure it'd be possible to find a suitable time for each attendee of a given workshop (these would be 5-6 hours altogether each I'd imagine), if people are willing to stay up late or get up early.

**Reply by duxonline** [reply]
Thank you for the consideration.

**Comment by deanmarcussen** [comment]
@Piedone yes, I generally think the workshops were very successful, however they where somewhat limited in the number of attendees, which was fine, but ultimately reduced the amount of value to the community, as we where only able to help 5-10 people (I can't remember the exact numbers).

There was quite a lot of discussion at the time around whether we should or should not record the sessions, and whether participants would be happy with that.

I personally feel that they would offer significantly more value if they where recorded, and able to be made available to the community afterwards.

Value was high however for users I think, so regardless, would be happy to run something again, though there is quite a lot of work in it, I think they're worthwhile

**Reply by Piedone** [reply]
Yeah, we intentionally capped the number of attendees at 10, with the goal of fostering a more interactive atmosphere. Not publishing the recordings was related to that too (so people would talk freely), but in the two workshops we provided, we did record them (with the permission of the attendees) and shared them with the attendees.

I'd still cap the attendee number for the same reason; not sure about not publishing the recording.

**Reply by deanmarcussen** [reply]
yeah makes sense.

I guess the thing I'm suggesting, or the issue I'm trying to solve, is how to reach a wider audience with them.

i.e. if you run them once a year for ten people, if that's the time that those ten people need to learn OC, great.
But it's not helping the person who needed to learn OC 11 months prior.

If that makes sense.

**Reply by Piedone** [reply]
Yeah, sure. I think if you want to learn OC on your own [the tutorials](https://docs.orchardcore.net/en/latest/docs/resources/tutorials/) are already available. I'd intend these workshops for those who instead want a workshop format.

I'd definitely do something more than once a year though, to keep up a good pace (and then we can see if there's actually enough demand).

**Reply by duxonline** [reply]
Ok, let me share my learning experience. I'm an experienced software engineer. I started with @Piedone's [video tutorial](https://www.youtube.com/playlist?list=PLuskKJW0FhJcSX7j0Bd-1X5hq3dgCtYwO). It took me a few months to finish it. I learned a lot and was feeling I wanted to write some modules.

I started converting one .Net Core application to OrchardCore module. The API/backend part went well. I was stuck on UI part because I didn't understand how Shape works or how I should use it. I had a whole bunch of questions like:
-  Should I convert all MVC views to shapes or only the ones that users may want to override?
-  Should I use `Factory.CreateAsync` or inherit from `ShapeViewModel`?
-  What's the best practicses of designing UI based on Shape? 
-  What are local Zones? Where are they defined or how it works? 
-  What is the best way to design a 3rd party module so that people can orverride it without changing its source code? 
-  Etc, etc

So I paused the project and tried to learn more about Shape or display management in general. I spent a few weekends debugging different Core modules and still didn't feel confident. Then I decided to take a look at [demo module tutorial](https://github.com/Lombiq/Orchard-Training-Demo-Module), that's when things started clicking. I was feeling more confident after I went through the sections on display management and content field display and editor.

With what I learned so far, I took another look at OrchardCore source code to gain more guidance. Only this time, I started with pull requests. Luckily I picked up @MikeAlhayek's work on [UserMenu](https://github.com/OrchardCMS/OrchardCore/pull/14214) and [SMS module](https://github.com/OrchardCMS/OrchardCore/pull/14125). After debugging the PRs multiple times, I finally was able to understand how a piece of functionality was implemented end to end. 

Now I'm comfortable starting my UI work again. Thank you @Piedone and @MikeAlhayek.

**Reply by Piedone** [reply]
I'm glad to hear, and thank you for sharing this!

What would have helped you between getting stuck with the UI and checking out shapes in the Training Demo module? Weren't the theming tutorials in the Dojo Course (starting with [this one](https://www.youtube.com/watch?v=yDtUyQpormU&list=PLuskKJW0FhJcSX7j0Bd-1X5hq3dgCtYwO&index=21)) suitable to get you started?

**Reply by duxonline** [reply]
To be honest, it's not enough for me. Somehow I didn't have a clear understanding as follows:

MvcController -> DisplayManager.BuildDisplayAsync -> DisplayDrivers.BuildDisplayAsync -> Controller View -> Shape Template (Book.cshtml) -> Child Shape Templates

I remember I was having an aha moment when I found the following code in `DisplayManager.BuildDisplayAsync`:
```            
  await _drivers.InvokeAsync(async (driver, model, context) =>
  {
      var result = await driver.BuildDisplayAsync(model, context);
      if (result != null)
      {
          await result.ApplyAsync(context);
      }
  }, model, context, _logger);
```
It's the missing piece for me.

**Reply by Piedone** [reply]
I see. For such in-depth, or specific questions, perhaps the best is a workshop.

**Comment by larremp** [comment]
Hi @Piedone , this is great discussion. If you need someone to organize these workshops, I'd like to volunteer. Maybe we can discuss in the next Teams meeting. Thanks and more power to this community!

**Reply by Piedone** [reply]
Thank you for the offer, would be great! (For those who don't know: Larrem organized the original workshops too.)

I'd like to see who can give workshops if not from the Lombiq team (see https://github.com/OrchardCMS/OrchardCore/discussions/14449#discussioncomment-7383931), and once we have that, then immediately get the word out there, publish details, and see who signs up. So, at that point, it would be awesome if you could help!

**Comment by Piedone** [comment]
In the meantime, we discussed this with the Lombiq team, to see who'd be able to give workshops. We can cover all the topics if nobody else wants to. So, I'd ask again, is there anybody who'd like to give one or more workshop, and preferably do this repeatedly?

**Reply by hishamco** [reply]
> We can cover all the topics if nobody else wants to.

Count me in, and don't forget to send me a Lombiq shirt :)

**Reply by Piedone** [reply]
You still need to send your postal address for that though :D.

**Comment by larremp** [comment]
Well, I'd like to suggest one workshop and I'll volunteer to teach it for free. Please see below my initial thoughts on the target audience, the objectives and topics. Of course, as we progressed, some things below might change.

Orchard Core 101 - Basics To Get You Started
Workshop Fee - Free
Duration - Maybe 1 - 2 hours

Target Audience

1. Folks who have full-time jobs working on [ASP.NET](http://asp.net/) MVC projects in their companies and want to explore [ASP.NET](http://asp.net/)-based CMSes.
2. Folks who are experienced in [ASP.NET](http://asp.net/) MVC but not yet very familiar with [ASP.NET](http://asp.net/) Core and want to explore [ASP.NET](http://asp.net/)-based CMSes.
3. Folks who had some previous experience with Orchard and want to dive-in immediately.

Objectives:
1. Get started immediately with a local development environment and a site on the web.
2. Remove  initial technical difficulties and frustration in setting up something unfamiliar. 
3. NOT meant to explain the details of Orchard Core, but to try to get them started in their Orchard Core journey.

Topics
1. How to get the stable version of OC?
2. How to create a development environment for Windows 11?
3. How to use GitHub as a respository?
4. How to publish in Azure using App Services?
5. How to have a separate App_Data folder for local development and for Azure?
6. How to use with Azure’s SQL databases?
7. How to use Azure’s storage account for images?
8. How to create your own Content Types?
9. How to enable and use Search?
10. How to enable and use SQL apis?
11. How to create a child Theme?
12. How to use Recipes to backup and import OC items?
13. How to upgrade to a new version?
14. How to get involved in the community?

**Reply by Piedone** [reply]
Sounds awesome, and such a workshop would be a great addition. Some notes about the content:

I think this is a bit too much for even 2 hours. I'd remove the Azure-related parts apart from perhaps some mentions about hosting options since you can't really fit enough into this time to be detailed apart from publishing from VS or something like that.

**Reply by larremp** [reply]
Thanks for the feedback, @Piedone. For now, I'd leave it there for the following reasons:

1. The audience I'm trying to reach are folks who need to be able to show to their bosses/supervisors a "proof-of-concept" quickly to get executive support. Hence, if we can show them how easy it is to publish a site, that would be great.
2. These folks are also working for companies that have tight integration with Azure, so if we can show them how easy it is to use storage accounts and NOT populate their web servers with images, that would be another plus.

I might be wrong, though, and so I'll adjust as we get going. But as they say, "Better be overdressed on the first date and secure the next." 😃

**Reply by larremp** [reply]
Just a few more additional comments:

1. I'm trying to reach small and medium sized companies who have legacy ASP.NET or even VB applications and are exploring technologies on where to migrate these apps.
2. Based on what I've seen, most of these folks think have either WordPress or Drupal as "top-of-mind" when they are exploring.
3. I think these folks are the "low hanging fruits" we can easily convince to try out Orchard Core because of their affinity with Microsoft tools and Azure.
4. We just have to make it easier for them to jump in and start their journey.

Cheers and more power to the Orchard Core community!

**Reply by Piedone** [reply]
Those people are definitely a target audience to grow the Orchard community from.

I just think that even if you do 2 hours, you won't be able to fit enough details and Q&A in there to allow people to do the same on their own. If you just want to demo it with the goal of "hey, this is something that you can also do, but don't worry about the details right now" then that's fine.

**Reply by larremp** [reply]
I think you might be right, @Piedone. Maybe I'll make it 2-3 hours. There might be one more topic I'll add, but I need to sort out the details first. Thanks.

**Comment by rwxzig** [comment]
Please do, there is so much potential to OrchardCore, and please consider doing live streams on YouTube. I also would suggest creating a list of topics that developers are struggling the most, and perhaps prioritize doing some tutorial about topics that are higher on priority.

**Reply by hishamco** [reply]
> and perhaps prioritize doing some tutorial about topics that are higher on priority.

Please check [Lombiq channel](https://youtube.com/@lombiq/playlists), they did a great video content on both Orchard & Orchard Core

**Reply by Piedone** [reply]
Thanks for your feedback! Can you tell about the topics _you_ struggle with?

**Reply by rwxzig** [reply]
@hishamco I am one of the biggest fans of Lombiq channel and their blog.
@Piedone sorry for the late reply, I had to close my exams first.  Tbh I am struggling with the base concepts such as shapes, forms, parts, to a point that I was about to give up and just use Orchard as backend/cms and then use razor pages for frontend, like sebastien ross mentioned in some orchard meeting, but I gave it another try because I really like this framework and I will invest more time and effort to it.

**Reply by Piedone** [reply]
I see, thanks. Are you aware of the [Dojo Course 3](https://orcharddojo.net/orchard-training/dojo-course-3-the-full-orchard-core-tutorial)? Because it starts with such concepts, and covers the most important ones throughout, both in videos and in code.

BTW looking at your interest under your profile, you might like this project of ours :) https://github.com/Lombiq/Hastlayer-SDK.

**Comment by duxonline** [comment]
Any update on this?

**Reply by Piedone** [reply]
Not yet. Are you interested in attending, perhaps?

**Reply by duxonline** [reply]
Yes, I am interested.

**Reply by Piedone** [reply]
Awesome! I'll keep everyone updated here.

**Comment by blueberryfy** [comment]
Hi @Piedone, let's start to actively plan these workshops. I've put together a version 1 of the Orchard Core 101. Here's the [link](https://www.basicstostart.com/orchard-core) 

Feedback is greatly appreciated.

**Reply by hishamco** [reply]
Great work Larrem, appreciated!!

**Reply by larremp** [reply]
Thanks, @hishamco ! If you see anything in the content that can be improved, please don't hesitate to give feedback. Thanks!

**Reply by hishamco** [reply]
Sure

**Reply by rwxzig** [reply]
@larremp I don't know, maybe it is just me but if you could do something about theming, Shapes, Parts, Zones etc, Something that I found confusing is that sometimes you see `shape_remove_item` or other rendering content. I am a bit confused on what is being inherited and how the flow of the displaying works. I am spending quite some time practicing and seeing the console on what changes in different pages, and I am finally starting to grasp it, but it would be nice if someone would make it easy for newcomers. 
I think most people would like to know about theming, rather than complicated topics.

**Reply by larremp** [reply]
Thanks for the feedback, @cptzig ! I agree with what you said. I think Theming and Shapes might be Orchard Core 20x. I'm confident as a community we'll be able to put together something regarding this topic.

**Reply by duxonline** [reply]
@larremp I like the section on publishing Orchard Core to Azure. I will join the classs if time is right. I'm in UTC+11 (Sydney, Australia). I don't mind staying up late or getting up early a bit for it.

**Reply by duxonline** [reply]
@larremp In a few hours class, I would like to see it focuses on one area or two and go reasonably deep instead of covering many topics. For example, for the first class we may focus on getting site up and running, the second one on core concepts like content fileds/part/type/item, layout, themes, tenants, etc, and the third one on theme development and then module development, even advanced module development or deep dive into shapes after that.

**Reply by larremp** [reply]
Thanks for the feedback, @XinOnline ! As we get these workshops going, I'm confident as a community, we can improve and make these better.

**Reply by Piedone** [reply]
Ah, very nice, Larrem! I can't promise we'll have similarly detailed guides for the other workshops :). Did you follow what we discussed under https://github.com/OrchardCMS/OrchardCore/discussions/14449#discussioncomment-7384815?

And thank you for following up here, this matter wasn't in my focus as much as it ought to, I'll get back to it.

**Reply by larremp** [reply]
Ok, thanks, @Piedone. Whenever you're ready, let's discuss so we can start announce the initial schedules and start running these workshops.

Thanks.

**Reply by Piedone** [reply]
How about scheduling workshops like two weeks before Harvest, so newcomers can get into OC just before it? https://github.com/OrchardCMS/OrchardCore/discussions/15260

**Reply by hishamco** [reply]
Believe it or not, I was thinking about workshops yesterday, IMHO we can arrange them for June - July

**Reply by panayiotisp** [reply]
Are these still going to happen at some point?

**Reply by Piedone** [reply]
Still want to launch this, but didn't have the time for it.

**Reply by hishamco** [reply]
Is this organized by Lombiq or we can help you on this

**Reply by Piedone** [reply]
You can help :)

**Reply by hishamco** [reply]
@agriffard are you interested to join the workshop as a trainer?
